bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/location/BasicLocation.java

```diff
@@ -150,11 +161,25 @@ public synchronized boolean set(URL value, boolean lock, String lockFilePath) th
 		}
 		lockFile = file;
 		location = value;
-		if (property != null)
+		if (property != null) {
 			configValues.setConfiguration(property, location.toExternalForm());
+		}
+		updateUrl(serviceRegistration);
```

Comment:
The `set` method should be changed to not be synchronous and then all of the method body should be moved into a `synchronized (this) {}` block except the lines:

```java
updateUrl(serviceRegistration);
return lock;
```

Reply:
done